rewrite to rust

Doublonmousse opened the pull request:
Go from a wrapper to a C++ library to a rust native implementation

Doublonmousse committed:
rewrite to rust, dynamics and wobble logic

Doublonmousse committed:
remove na dependency (use f32 instead)

Doublonmousse committed:
add state modeler and prediction

Doublonmousse committed:
update example svg

Doublonmousse committed:
use `ModelerResult` for return values

Doublonmousse committed:
remove `KalmanPredictorParams`

Doublonmousse committed:
Create todo.md

Doublonmousse committed:
fmt

Doublonmousse committed:
more tests and fixes

Doublonmousse committed:
more testing and fixes

Doublonmousse committed:
debug tests

Doublonmousse committed:
reserve capacities for `vecs` and `vecdecque`

Doublonmousse committed:
`Copy` vs clone changes

Doublonmousse committed:
doc updates (with mathml)

Doublonmousse committed:
clippy and fmt fixes

Doublonmousse commented:
There's one test (corresponding to `WobbleSmoothed` in the original `ink_stroke_modeler` repo that fails). Haven't figured out yet what's happening there

Doublonmousse committed:
fixing wobble test

Doublonmousse committed:
fix `generate_output_up_nodelta` test

Doublonmousse committed:
Update todo.md

Doublonmousse commented:
Ok, it finally passes tests. Not all tests are reimplemented though. Notable things missings are written in the `todo.md` file

Doublonmousse committed:
fmt

flxzt commented:
It would be nice to have the steps for regenerating the docs documented or even put into a [justfile](https://just.systems/man/en/) / script. Ideally this would then be added to the docs CI so we ensure the hosted github pages cargo docs are always up-to-date. And the generated intermediary html/pdf files should probably be not checked in, only the source typst?

flxzt commented on src/impl_ds.rs:
```diff
@@ -0,0 +1,216 @@
+    }
+
+    pub fn near(self, other: ModelerResult) -> bool {
+        let tol = 1e-4;
```
Consider using [approx](https://docs.rs/approx/0.5.1/approx/), especially it's `assert_rel_eq` for comparing floats

flxzt commented on src/impl_ds.rs:
```diff
@@ -0,0 +1,216 @@
+/// with printed output for debug purposes
+pub fn compare_results(left: Vec<ModelerResult>, right: Vec<ModelerResult>) -> bool {
+    if left.len() != right.len() {
+        // debug prints
```
Better would be to pull in `tracing` and use `tracing::debug!` here.

flxzt commented on src/lib.rs:
```diff
@@ -1,549 +1,177 @@
+    /// wobble smoother structures
+    /// deque to hold events that are recent
+    /// to calculate a moving average
+    wobble_decque: VecDeque<WobbleSample>,
```
typo: `decque` -> `deque`

flxzt commented on src/impl_ds.rs:
```diff
@@ -0,0 +1,216 @@
+        }
+    }
+
+    // helper methods : superfluous if we make the params of the struct public
```
Let's get rid of all getters here, we need to adjust the usage in Rnote anyway

Doublonmousse replied:
Done on the ink stroke modeler but keeping this open as I have to adjust things on the rnote side as well

Doublonmousse replied:
Working on the rnote side

flxzt commented on src/impl_ds.rs:
The separation of struct definitions in `lib.rs` and the `impl` blocks here is unusual for rust code. I'd expect them to be in the same file. What about separating to `input.rs`, `result.rs`, `params.rs`, `modeler.rs` and maybe re-export in `lib.rs` to make them easy to use as the public API

flxzt commented on src/position_modeler.rs:
```diff
@@ -0,0 +1,754 @@
+/// This struct models the movement of the pen tip based on the laws of motion.
+/// The pen tip is represented as a mass, connected by a spring to a moving
+/// anchor; as the anchor moves, it drags the pen tip along behind it.
+pub struct PositionModeler {
```
I don't think this needs to be public. I'd say the modelers are internal implementation only. (Same for the other modeler and all fields and methods)

Doublonmousse replied:
I've changed the visibility, mostly to `pub(crate)`. Is it what you meant ?

flxzt commented on src/testing.rs:
Consider placing the unit test into the same files as the implementation code, that's the convention in most rust projects (and you already did that with other tests)

flxzt commented on src/utils.rs:
We should keep all functions that are internal only private

flxzt commented on src/state_modeler.rs:
```diff
@@ -0,0 +1,269 @@
+#[test]
+fn state_modeler_straight() {
+    let mut state_mod = StateModeler::new(10);
+    assert_eq!(state_mod.query((0.0, 0.0)), 1.0); // 1 is our "unknown" default value
```
It would be better to use `approx-eq`'s `assert_rel_eq` here as well and in all other occurrences of float compares

flxzt commented on src/testing.rs:
```diff
@@ -0,0 +1,2295 @@
+    }
+
+    #[test]
+    fn test_clamping() {
```
testing rust's `clamp` is not needed I think

flxzt commented on src/impl_ds.rs:
```diff
@@ -0,0 +1,216 @@
+    }
+}
+
+impl ModelerResult {
```
Same here, let's just use the fields directly instead of keeping the getters

flxzt reviewed (commented):
I only did a high level review for now with things that I immediately noticed. The modeler itself seems to be well tested and working fine, so I'll only quickly go over it in a second pass

Doublonmousse committed:
justfile for documentation

Doublonmousse committed:
minor fixes

- use`tracing::debug`
- typo for decque
- minor docstring changes
- remove the rust test of clamp for floats

Doublonmousse committed:
make testing and utils private

Doublonmousse committed:
visibility changes (pub(crate) instead of public)

- visibility changes
- re add back docs for CI to work

Doublonmousse committed:
additional visibility changes and fmt

Doublonmousse committed:
remove getters for `ModelerInput` and `ModelerResult`

Doublonmousse committed:
use `approx` for float comparison

Doublonmousse committed:
reorganise files (impl closer to their structs)

Doublonmousse committed:
reorganise tests

flxzt commented on src/engine.rs:
```diff
@@ -0,0 +1,2625 @@
+}
+
+#[cfg(test)]
+mod ink_stroke_modeler {
```
I'd name this module `tests`

flxzt commented on Cargo.toml:
```diff
@@ -4,19 +4,11 @@ license = "MIT OR Apache-2.0"
 svg = "0.16.0"
+tracing = "0.1.40"
+
+[dependencies]
```
No dependencies - nice!

flxzt commented on src/engine.rs:
```diff
@@ -0,0 +1,2625 @@
+    ///
+    /// Here the error is also obsolete as the `ModelerParams` is expected
+    /// to have been built with [ModelerParams::validate] that validates the
+    /// parameters
```
I think validating internally is fine, so I would suggest removing this comment and keep the `Result`. However, the `i32` error variant does not make sense anymore.
Let's change it to `String`, refactoring to proper error types can be done later

flxzt commented on src/engine.rs:
```diff
@@ -0,0 +1,2625 @@
+
+    /// Clears any in-progress stroke, keeping the same model parameters
+    /// Technically the error is not needed anymore as the params are kept
+    /// as is and can't be uninitialized
```
Then let's just remove returning `Result` entirely

flxzt commented on src/engine.rs:
```diff
@@ -0,0 +1,2625 @@
+    ///
+    /// for now rnote's wrapper codes verify that the input is not duplicated and time increases between strokes
+    /// This is not tested here, as we suppose that these things are verified beforehand
+    pub fn update(&mut self, input: ModelerInput) -> Result<Vec<ModelerResult>, i32> {
```
Let's change the error variant to String here as well

flxzt commented on src/engine.rs:
```diff
@@ -0,0 +1,2625 @@
+    ///position and the raw position based on the speed.
+    ///high speeds movements won't be smoothed but low speed will.
+    #[doc = include_str!("../docs/wobble.html")]
+    pub fn wobble_update(&mut self, event: &ModelerInput) -> (f32, f32) {
```
Probably not intended to be public?

flxzt commented on src/engine.rs:
```diff
@@ -0,0 +1,2625 @@
+    /// If this does not return an error, results will contain at least one Result, and potentially
+    /// more if the inputs are slower than the minimum output rate
+    ///
+    /// for now rnote's wrapper codes verify that the input is not duplicated and time increases between strokes
```
We should probably move the validation into ink-stroke-modeler-rs and remove the mention of Rnote, making it easier to use the the library for other projects

flxzt commented on src/engine.rs:
```diff
@@ -0,0 +1,2625 @@
+                self.wobble_update(&input); // first event is "as is"
+
+                // create the position modeler
+                self.position_modeler = Some(PositionModeler::new(self.params, input.clone()));
```
Can there be a check that the first event is `Down`, because otherwise users only get a unexplained panic when the modeler is unwrapped and it might not be clear what the issue is.

Doublonmousse replied:
I've added this in the error validation part.
This means that an error occurs if a Down event occurs whilst a stroke is in progress.

flxzt commented on src/position_modeler.rs:
```diff
@@ -0,0 +1,749 @@
+}
+
+impl ModelerPartial {
+    #[allow(dead_code)]
```
`allow(dead_code)` not needed I think

flxzt commented on src/input.rs:
```diff
@@ -0,0 +1,34 @@
+#[allow(unused)]
+pub enum ModelerInputEventType {
+    /// For the first pen down event (pen touches the screen)
+    kDown,
```
I don't know why the cpp `ink-stroke-modeler` lib named these `k*`, we don't have to follow that anymore. We can name them `Down`, `Move`, `Up`

Doublonmousse replied:
Done but needing some API changes on the note side that I'll do a little later

flxzt reviewed (commented):
I am currently getting a panic when trying to run the stroke example:

```
thread 'main' panicked at library/alloc/src/raw_vec.rs:25:5:
capacity overflow
stack backtrace:
   0: rust_begin_unwind
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/panicking.rs:72:14
   2: alloc::raw_vec::capacity_overflow
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/alloc/src/raw_vec.rs:25:5
   3: alloc::raw_vec::handle_reserve
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/alloc/src/raw_vec.rs:566:34
   4: alloc::raw_vec::RawVec<T,A>::with_capacity_in
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/alloc/src/raw_vec.rs:155:9
   5: alloc::vec::Vec<T,A>::with_capacity_in
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/alloc/src/vec/mod.rs:688:20
   6: alloc::vec::Vec<T>::with_capacity
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/alloc/src/vec/mod.rs:481:9
   7: ink_stroke_modeler_rs::engine::StrokeModeler::update
             at ./src/engine.rs:242:35
   8: stroke::main::{{closure}}
             at ./examples/stroke.rs:97:13
   9: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/ops/function.rs:294:13
  10: core::iter::traits::iterator::Iterator::find_map::check::{{closure}}
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/iter/traits/iterator.rs:2905:32
  11: core::iter::traits::iterator::Iterator::try_fold
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/iter/traits/iterator.rs:2409:21
  12: core::iter::traits::iterator::Iterator::find_map
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/iter/traits/iterator.rs:2911:9
  13: <core::iter::adapters::filter_map::FilterMap<I,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/iter/adapters/filter_map.rs:64:9
  14: <core::iter::adapters::fuse::Fuse<I> as core::iter::adapters::fuse::FuseImpl<I>>::next
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/iter/adapters/fuse.rs:359:9
  15: <core::iter::adapters::fuse::Fuse<I> as core::iter::traits::iterator::Iterator>::next
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/iter/adapters/fuse.rs:50:9
  16: <core::iter::adapters::flatten::FlattenCompat<I,U> as core::iter::traits::iterator::Iterator>::next
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/iter/adapters/flatten.rs:610:29
  17: <core::iter::adapters::flatten::Flatten<I> as core::iter::traits::iterator::Iterator>::next
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/iter/adapters/flatten.rs:298:9
  18: alloc::vec::Vec<T,A>::extend_desugared
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/alloc/src/vec/mod.rs:2988:35
  19: <alloc::vec::Vec<T,A> as alloc::vec::spec_extend::SpecExtend<T,I>>::spec_extend
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/alloc/src/vec/spec_extend.rs:17:9
  20: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/alloc/src/vec/spec_from_iter_nested.rs:43:9
  21: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/alloc/src/vec/spec_from_iter.rs:33:9
  22: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/alloc/src/vec/mod.rs:2894:9
  23: core::iter::traits::iterator::Iterator::collect
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/iter/traits/iterator.rs:2003:9
  24: stroke::main
             at ./examples/stroke.rs:94:25
  25: core::ops::function::FnOnce::call_once
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Doublonmousse commented:
Okay, found out the reason for the panic : that was me messing up the API change and inverting pressure and time fields resulting in inputs that aren't in chronological order.

Might as well add checks on the time parameters in the modeler to catch this

Doublonmousse committed:
fix stroke example

Doublonmousse committed:
renamings

- rename module for tests
- remove `k` from variants
- removing result from `reset` (to verify)

Doublonmousse committed:
fixes for reset and better validation

Doublonmousse committed:
reset the state modeler and use the new params when resetting with parameters

Doublonmousse committed:
error if two down event occur in the same stroke

Doublonmousse committed:
removing the last k

Doublonmousse committed:
catch duplicate event and negative time delta

The stroke modeler is left intact through this, making it possible to redo predictions as if the input was ignored

Doublonmousse committed:
fmt and doc update

Doublonmousse committed:
fix reset bugs

flxzt commented:
From a code structure perspective this looks good to me now. Do you want to complete the rest of the tasks in the todo here or should they be done at a later time?

One other thing left to do: the README needs an update, mentioning that `ink-stroke-modeler-rs` is a rust reimplementation now.

Doublonmousse committed:
Update todo.md

Doublonmousse committed:
Update README.md

Doublonmousse commented:
I don't know. I think I can do them now relatively quickly.
- changing settings might be done later (cause that'd be dependent on the document opened and dpi settings) and that might not be quick
- API changes would be to change everything to f64 and use rust durations for times to have less conversions going on between rnote and the modeler
- change error types to be strongly typed (at least in the ink stroke modeler and maybe later changing the code in the rnote side)

Doublonmousse committed:
strongly typed errors for the `update` calls

Doublonmousse committed:
Merge branch 'riir' of https://github.com/Doublonmousse/ink-stroke-modeler-rs into riir

Doublonmousse committed:
export errors

Doublonmousse committed:
test (change to f64)

Doublonmousse commented:
I'm finishing up changing the error types for the `update` calls only, the rest will probably be in a follow-up (for the API change, going from f32 to f64 seems fine although some tests fail, and I need to double check that I haven't done anything weird and come up with new tests if these fails are f32/f64 related)

Doublonmousse commented:
Ok, I'll stop there for today
- I've changed the readme as you asked
- I've changed the error types for the `update` calls and changed the relevant code in the rnote part.

I'll do the rest of the things as a follow-up

Doublonmousse committed:
change tolerance for f64

Doublonmousse committed:
update todo

Doublonmousse committed:
Merge pull request #1 from Doublonmousse/f64_change

F64 change

flxzt commented:
Alright, the change to f64 also looks good to me.

I pushed one last improvement for the cpp-bindings and created the archive branch `archive-cpp-bindings`.

Once the new conflicts are resolved I'll squash-merge `riir`

Doublonmousse committed:
Merge branch 'riir' of https://github.com/Doublonmousse/ink-stroke-modeler-rs into Doublonmousse-riir

Doublonmousse committed:
Update Cargo.toml

Doublonmousse committed:
Update README.md

Doublonmousse committed:
Merge pull request #2 from flxzt/Doublonmousse-riir

resolve merge conflicts

Doublonmousse commented:
Okay, conflicts are finally resolved

Doublonmousse committed:
remove exported submodules

Doublonmousse committed:
remove submodules

flxzt commented:
Awesome. Thank you so much again for doing this huge amount of work. :)
